AlgebraInLean/World0LeanIntroSOLUTIONS/Sheet2Solutions.lean

```diff
@@ -0,0 +1,283 @@
+example : P → (P → Q) → Q := by
+  intro h1 h2
+  apply h2 at h1
+  exact h1
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q) → (Q → R) → P → R := by
+  intro h1 h2 h3
+  apply h1 at h3
+  apply h2 at h3
+  exact h3
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q → R) → (P → Q) → P → R := by
+  intro h1 h2 h3
+  apply h1
+  exact h3
+  apply h2 at h3
+  exact h3
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → R) → (S → Q) → (R → T) → (Q → R) → S → T := by
+  intro h1 h2 h3 h4 h5
+  apply h2 at h5
+  apply h4 at h5
+  apply h3 at h5
+  exact h5
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (P → Q) → ((P → Q) → P) → Q := by
+  intro h1 h2
+  apply h1
+  apply h2
+  exact h1
+  done
+
+/-FM Section 1 Sheet 1-/
+example : ((P → Q) → R) → ((Q → R) → P) → ((R → P) → Q) → P := by
+  intro h1 h2 h3
+  apply h2
+  intro h4
+  apply h1
+  intro h5
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : ((Q → P) → P) → (Q → R) → (R → P) → P := by
+  intro h1 h2 h3
+  apply h1
+  intro h4
+  apply h3
+  apply h2
+  exact h4
+  done
+
+/-FM Section 1 Sheet 1-/
+example : (((P → Q) → Q) → Q) → P → Q := by
+  intro h1 h2
+  apply h1
+  intro h3
+  apply h3
+  exact h2
+  done
+
+/-FM Section 1 Sheet 1-/
+example :
+    (((P → Q → Q) → (P → Q) → Q) → R) →
+      ((((P → P) → Q) → P → P → Q) → R) → (((P → P → Q) → (P → P) → Q) → R) → R := by
+  intro h1 h2 h3
+  apply h2
+  intro h4 h5 h6
+  apply h4
+  intro h7
+  exact h5
+  done
+
+/-The ↔ means "if and only if." If you are ever curious how to type a certain symbol in Lean, just
+hover over it for a few seconds. rw can be applied to ↔ hypotheses.-/
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ P) := by
+  intro h
+  rw [h]
+  done
+
+/-FM Section 1 Sheet 5-/
+example : (P ↔ Q) → (Q ↔ R) → (P ↔ R) := by
+  intro h1 h2
+  rw [h2] at h1
+  exact h1
+  done
+
+/-Note that many of the exercises above could be solved in one line, using the tactic "tauto," just
+as some levels in NNG could be solved by "tauto."-/
+
+/-Lean also has booleans True and False.
+
+If the goal is "True", it can be solved with the tactic "trivial".
+
+For example:-/
+
+/-FM Section 1 Sheet 2-/
+example : True := by
+  trivial
+  done
+
+/-Let's do some exercises, adding "trivial" to our list of tactics.-/
+
+/-FM Section 1 Sheet 2-/
+example : True → True := by
+  intro h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : False → True := by
+  intro h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : False → False := by
+  intro h1
+  exact h1
+  done
+
+/-FM Section 1 Sheet 2-/
+example : (True → False) → False := by
+  intro h1
+  apply h1
+  trivial
+  done
+
+/-FM Section 1 Sheet 2-/
+example : True → False → True → False → True → False := by
+  intro h1 h2 h3 h4 h5
+  exact h2
+  done
+
+/-FM Section 1 Sheet 2-/
+example : P → (P → False) → False := by
+  intro h1 h2
+  apply h2 at h1
+  exact h1
+  done
+
+/-The tactic "exfalso" changes any goal after the ⊢ symbol to "False."
```

Comment:
Why is it called `exfalso`? Good opportunity to note a quirk of Lean logic: if you have a proof of False, you can generate a proof of anything.